D8 initial mapping

achaikou opened the pull request.

achaikou committed:
Reorder declarations in segy.h

Declaration order matters in C, so it makes more sense to reorder
existing entities as enums - structs - function declarations.

achaikou committed:
Fix check in segy_sorting

Used fields are 1-based, yet they are checked as if they were 0-based.

achaikou committed:
Inline fd_get_int/fd_set_int

With more get-set related functions coming, inlining those two would
make reading code more manageable.

achaikou committed:
Add binary/trace wrappers for get/set field

We in essence bring back segy_get/set_(b)field functions.
With this to switch from previous major version to new one, users would
just have to replace names like segy_get_bfield with
segy_get_binfield_int.

Wrappers allow us to keep user interface mostly unchanged but still
nicely incorporate coming mapping changes.

achaikou committed:
Switch lookup tables to D8 compatible mappings

D8 introduces new functionality to SEG-Y specification. Now field
mappings for trace headers could be defined by the user, per file.

We need to update our own lookup tables to be compatible with that
mapping structure.

Those tables will become a default layout for parsing traceheaders.
Binary tables are updated accordingly as we wish to treat them no
differently from trace tables.

Co-authored-by: Yngve Villanger <yvi@equinor.com>
Co-authored-by: Alexander Jaust <18347253+ajaust@users.noreply.github.com>

ajaust reviewed (commented):
Great job. The changes look good to me. 👍 

I don't have anything serious. I think half of my comments are about `const`ness (which is negotiable) and the rest is for clarifications/better understanding.

yngve793 reviewed (commented):
Nice and structured commits that is pulling us in the right direction. :-)

My comments are generally about structure and how to organise data. No surprise there. :-) Now that we have real commits to start the discussion from, it should be easier to avoid talking past each other. Clearly not everything can be changed at once and some steps are intermediate. Let us discuss a couple of these comments at the next meting.

achaikou committed:
Make layout mapping a parameter

Prepare for traceheader mapping no longer being hardcoded.

Public traceheader get/set functions will require header mapping as a
parameter, binheader functions will use known one. All _int legacy
function will operate only on default headers.

For the moment all calls inside the library use int version and default
traceheader map.
It would be changed later when code is updated to support non-default
mappings.

Co-authored-by: Yngve Villanger <yvi@equinor.com>
Co-authored-by: Alexander Jaust <18347253+ajaust@users.noreply.github.com>

achaikou committed:
Make offset in internal functions 0-based

0-based index is something usually expected by the developer.
Get/set field functions become more readable and less error-prone when
everything - header, mapping and offset is unified and is 0-based.

Because users are used to trace fields being 1-based and binary fields
being 3201-based we decided against changing it in public facing
functions as well. It costs us some logical repetitions, but it is
acceptable sacrifice.

achaikou reviewed (commented):
Thanks for the review guys! :smiley:

ajaust reviewed (approved):
Thanks for the replies and changes. I have no further comments.
